compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -2105,51 +2105,52 @@ static LogicalResult setVMVXRootConfigImpl(func::FuncOp entryPointFn,
-      ArrayRef<int64_t> innerTiles = packOp.getStaticInnerTiles();
-      ArrayRef<int64_t> dimPos = packOp.getInnerDimsPos();
-      for (auto [pos, size] : llvm::zip_equal(dimPos, innerTiles)) {
-        if (tileSizes[pos] == 0 || ShapedType::isDynamic(size))
-          continue;
-        tileSizes[pos] = llvm::alignTo(tileSizes[pos], size);
-        LLVM_DEBUG(KD_DBGS() << "Align # " << pos << " tile size to "
-                             << tileSizes[pos] << "\n");
-      }
-    }
-
-    return WalkResult::advance();
-  });
-  if (res.wasInterrupted())
+static LogicalResult
+adjustTileSizesForPackOp(func::FuncOp entryPointFn, tensor::PackOp packOp,
+                         SmallVector<int64_t> &distTileSizes,
+                         SmallVector<int64_t> &parallelVecTileSizes) {
+  // Align the tile sizes of the root op to the pack op's inner tile sizes, so
+  // we can derive the outer tile sizes for pack ops later in
+  // setLoweringConfigForComputeOps by dividing with inner tile sizes.
+  ArrayRef<int64_t> innerDimsPos = packOp.getInnerDimsPos();
+  ArrayRef<int64_t> innerTiles = packOp.getStaticInnerTiles();
+  // Currently we only handle pack op with static inner tile sizes.
+  if (llvm::any_of(innerTiles,
+                   [](int64_t size) { return ShapedType::isDynamic(size); })) {
     return failure();
-
-  auto pipeline = getTranslationInfo(entryPointFn).getPassPipeline().getValue();
-  return setOpConfigAndEntryPointFnTranslation(
-      entryPointFn, rootOp, tileSizesList,
-      loweringConfig.getScalableTileFlagVals(), pipeline);
+  }
+  // Pack op requires special vector tile sizes to achieve good performance.
+  // Override the parallel vector tile sizes from pack op.
+  auto vecTileSizes = getPackVectorTileSizes(entryPointFn, packOp);
+  auto outerDimsPerm = packOp.getOuterDimsPerm();
+  if (!outerDimsPerm.empty()) {
+    auto invertedPerm = invertPermutationVector(outerDimsPerm);
+    applyPermutationToVector(vecTileSizes, invertedPerm);
+  }
+  // Scale to actual tile sizes with the pack op's inner tile sizes.
+  for (auto [pos, size] : llvm::zip_equal(innerDimsPos, innerTiles)) {
+    vecTileSizes[pos] *= size;
+  }
+  for (auto [pos, size] : llvm::enumerate(vecTileSizes)) {
+    if (!size)
+      continue;
+    if (!parallelVecTileSizes[pos]) {
+      parallelVecTileSizes[pos] = size;
+      continue;
+    }
+    // If other ops already set a smaller tile size, don't override it to avoid
+    // too large tile size on them.
+    parallelVecTileSizes[pos] = std::min(parallelVecTileSizes[pos], size);
+  }
+  // Ensure tile sizes are aligned to the inner tile sizes.
+  for (auto [pos, size] : llvm::zip_equal(innerDimsPos, innerTiles)) {
+    if (distTileSizes[pos])
+      distTileSizes[pos] = llvm::alignTo(distTileSizes[pos], size);
+    if (parallelVecTileSizes[pos])
+      parallelVecTileSizes[pos] =
+          llvm::alignTo(parallelVecTileSizes[pos], size);
+  }
+  return success();
```

Comment:
I think having an example for the different cases or perhaps on the function header would be very helpful to follow what is going on here

Reply:
Done.